File: 121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if (n < 2) return 0;
        int mins[n];
        mins[0] = 100001;
        mins[1] = prices[0];
        
        for(int i = 2; i<n; ++i){
            mins[i] = min(prices[i-1], mins[i-1]);
        }
        int maxP = 0; 
        for(int i = 1; i<n; ++i){
            maxP = max(maxP, prices[i]-mins[i]);
        }
        return maxP;
    }
};

File: 121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        min_so_far = prices[0]
        max_diff = 0

        for i in range(len(prices)):
            diff = prices[i] - min_so_far
            if diff > max_diff:
                max_diff = diff                
            elif prices[i] < min_so_far:
                min_so_far = prices[i]

        return max_diff